### PR TITLE
Elaborate registry names to index URLs when publishing

### LINKED ISSUE #4880: Alternative registry references are inconsistent between the name and index URL
Dependency information in `Cargo.toml` refers to alternative registries by their name, while the dependency information in the index JSON refers to alternative registries by URL. We should consistently use one or the other.

@withoutboats

### PR DESCRIPTION
**Pull request by sfackler**
This avoids introducing a dependency on the publisher's name for the
registry.

Closes #4880

### CODE REVIEW

File: src/cargo/util/toml/mod.rs [hunk 1]
```diff
@@ -181,6 +181,7 @@ impl<'de> de::Deserialize<'de> for TomlDependency {
 pub struct DetailedTomlDependency {
     version: Option<String>,
     registry: Option<String>,
+    registry_index: Option<String>,
```
Comment (alexcrichton) [inline src/cargo/util/toml/mod.rs]: Could this use `#[serde(rename)]` to use `registry-index` instead?

File: src/cargo/util/toml/mod.rs [hunk 2]
```diff
@@ -925,12 +938,21 @@ impl TomlDependency {
...
 
-        let new_source_id = match (details.git.as_ref(), details.path.as_ref(), details.registry.as_ref()) {
-            (Some(_), _, Some(_)) => bail!("dependency ({}) specification is ambiguous. \
+        let new_source_id = match (
+            details.git.as_ref(),
+            details.path.as_ref(),
+            details.registry.as_ref(),
+            details.registry_index.as_ref(),
+        ) {
+            (Some(_), _, Some(_), _) |
```
Comment (alexcrichton) [inline src/cargo/util/toml/mod.rs]: One day we will make this better. It is not this day.

**Comment by sfackler** [comment]
Updated

File: src/cargo/util/toml/mod.rs [hunk 3]
```diff
@@ -178,25 +178,26 @@ impl<'de> de::Deserialize<'de> for TomlDependency {
 }
 
 #[derive(Deserialize, Serialize, Clone, Debug, Default)]
+#[serde(rename_all = "kebab-case")]
```
Comment (alexcrichton) [inline src/cargo/util/toml/mod.rs]: Ah unfortuantely I don't think we can do this as it breaks manifests that use `default_features = false`, for example.

When we first moved to serde we ran into that (as the old rustc-serialize toml-rs accepted both) and that's why there's two entries below :(

Comment (sfackler) [reply src/cargo/util/toml/mod.rs]: The `#[serde(rename = "default_features")]` down on default_features2 will handle that: https://play.rust-lang.org/?gist=db44b7c463a8b4ac43c0b67e2b36e5d0&version=stable

Comment (alexcrichton) [reply src/cargo/util/toml/mod.rs]: Oh! I missed that, nice!

**Comment by alexcrichton** [comment]
This LGTM, but others from @rust-lang/cargo will likely wish to weigh in

**Comment by carols10cents** [comment]
One reason we didn't want to support this is to disallow credentials (such as https://username:password@index-url.com) from getting into the Cargo.toml. We could just always explicitly ignore credentials specified here so that doing that doesn't work.....

**Comment by sfackler** [comment]
Should we forbid credentials in index URLs entirely?

**Comment by sfackler** [comment]
Forbidding them would also take care of potential issues if some people use `ssh://foobar.com` and others use `ssh://git@foobar.com`, for example.

**Comment by alexcrichton** [comment]
I think so long as we have some way of specifying credentials in `~/.cargo/config` it makes sense to forbid credentials in the manifest itself (although maybe allow a username as existing git/ssh deps may use that). Not that I think we should require the `~/.cargo/config` configuration to exist before landing this of course!

**Comment by sfackler** [comment]
Cool, so are there any blockers on this PR, then?

**Comment by sfackler** [comment]
Updated to forbid credentials in registry URLs.

**Pull request merged**